usr/lib/common/key.c

```diff
@@ -3032,6 +3038,240 @@ CK_RV ibm_dilithium_priv_unwrap(TEMPLATE *tmpl, CK_BYTE *data,
+    CK_RV rc;
+
+    rc = ber_decode_IBM_KyberPublicKey(data, total_length, &pk,
+                                       &value, &oid);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ber_decode_IBM_KyberPublicKey failed\n");
+        return rc;
+    }
+
+    rc = ibm_pqc_add_keyform_mode(tmpl, oid, CKM_IBM_KYBER);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ibm_pqc_add_keyform_mode failed\n");
+        return rc;
+    }
+
+    rc = template_update_attribute(tmpl, pk);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("template_update_attribute failed.\n");
+        goto error;
+    }
+    pk = NULL;
+    if (add_value) {
+        rc = template_update_attribute(tmpl, value);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("template_update_attribute failed.\n");
+            goto error;
+        }
+    } else {
+        free(value);
+    }
+    value = NULL;
+
+    return CKR_OK;
+
+error:
+    if (pk)
+        free(pk);
+    if (value)
+        free(value);
+
+    return rc;
+}
+
+//
+//
+CK_RV ibm_kyber_priv_unwrap(TEMPLATE *tmpl, CK_BYTE *data,
+                            CK_ULONG total_length, CK_BBOOL add_value)
+{
+    CK_ATTRIBUTE *sk = NULL, *pk = NULL, *value = NULL;
+    const struct pqc_oid *oid;
+    CK_RV rc;
+
+    rc = ber_decode_IBM_KyberPrivateKey(data, total_length,
+                                        &sk, &pk, &value, &oid);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ber_decode_IBM_KyberPrivateKey failed\n");
+        return rc;
+    }
+
+    rc = ibm_pqc_add_keyform_mode(tmpl, oid, CKM_IBM_KYBER);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ibm_pqc_add_keyform_mode failed\n");
+        return rc;
+    }
+
+    rc = template_update_attribute(tmpl, sk);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed\n");
+        goto error;
+    }
+    sk = NULL;
+    rc = template_update_attribute(tmpl, pk);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed\n");
+        goto error;
+    }
+    pk = NULL;
+    if (add_value) {
+        rc = template_update_attribute(tmpl, value);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("template_update_attribute failed.\n");
+            goto error;
+        }
+    } else {
+        free(value);
+    }
+    value = NULL;
+
+    return CKR_OK;
+
+error:
+    if (sk)
+        free(sk);
+    if (pk)
+        free(pk);
+    if (value)
+        free(value);
+
+    return rc;
+}
+
+CK_RV ibm_pqc_publ_get_spki(TEMPLATE *tmpl, CK_KEY_TYPE keytype,
+                            CK_BBOOL length_only,
+                            CK_BYTE **data, CK_ULONG *data_len)
+{
+    switch (keytype) {
+    case CKK_IBM_PQC_DILITHIUM:
+        return ibm_dilithium_publ_get_spki(tmpl, length_only, data, data_len);
+    case CKK_IBM_PQC_KYBER:
+        return ibm_kyber_publ_get_spki(tmpl, length_only, data, data_len);
+    default:
+        TRACE_DEVEL("Kytype 0x%lx not supported.\n", keytype);
+        return CKR_KEY_TYPE_INCONSISTENT;
```

Comment:
Hmm interesting ... in a previous commit there was a similar case of a key type not being supported. There you returned CKR_MECHANISM_INVALID (in policy.c line 207/208). I searched for a "key type unsupported" rc, but such an rc does not exist. Here you decided to use CKR_KEY_TYPE_INCONSISTENT, which is not exactly what we would need ;-)

Reply:
well, the caller of this function expects to deal with a key type of Dilithium or Kyber, so if it is not one of that, the key type is inconsistent with the operation performed. CKR_MECHANISM_INVALID would be totally wrong here. 

Yes for policy.c line 207/208 it should probably also be CKR_KEY_TYPE_INCONSISTENT.